smart-contracts/contracts/cl-vault/src/vault/deposit.rs

```diff
@@ -36,90 +41,218 @@ pub(crate) fn _execute_any_deposit(
+
+    // create a position for each position
+    let msgs: Result<Result<Vec<SubMsg>, ContractError>, ContractError> = position_funds
+        .map(|((pos1, amount0, amount1), (pos2, fp))| {
+            if pos1.position_id != pos2.position_id {
+                // this would be a programming error
+                panic!("Position ids in the zipped iterator don't match")
+            }
+
+            // we want to make sure deposit 1 is item 1 in the queue, deposit 2 item 2 in the queue
+            // TODO can this be less complex?
+            CURRENT_DEPOSIT.push_back(
+                deps.storage,
+                &CurrentDeposit {
+                    token0_in: *amount0,
+                    token1_in: *amount1,
+                    refund0: None,
+                    refund1: None,
+                    liquidity_out: None,
+                    original_id: pos1.position_id,
+                },
+            )?;
+
+            total_allocated = (total_allocated.0 + *amount0, total_allocated.1 + amount1);
+
+            create_position(
+                deps.branch(),
+                &env,
+                fp.position.as_ref().unwrap().lower_tick,
+                fp.position.unwrap().upper_tick,
+                CoinList::from_coins(vec![
+                    coin(amount0.u128(), &token0.denom),
+                    coin(amount1.u128(), &token1.denom),
+                ])
+                .coins(),
+                Uint128::zero(),
+                Uint128::zero(),
+            )
+        })
+        .map(|msg| {
+            msg.map(|m| -> Result<SubMsg<_>, ContractError> {
+                Ok(SubMsg::reply_on_success(
+                    m,
+                    Replies::DepositCreatePosition as u64,
+                ))
+            })
+        })
+        .collect();
 
-    let create_position_msg = create_position(
-        deps,
-        &env,
-        position.lower_tick,
-        position.upper_tick,
-        sort_tokens(coins_to_send),
-        Uint128::zero(),
-        Uint128::zero(),
+    // create the mint callback message, and set that after the create_position msges, such that after we have created all positions for the user,
+    // we can figure out how to mint by the state of CURRENT_DEPOSIT
+    let mint = ExecuteMsg::VaultExtension(crate::msg::ExtensionExecuteMsg::CallbackExecuteMsg(
+        crate::msg::CallbackExecuteMsg::MintUserDeposit {},
+    ));
+
+    // TODO save any funds we did not use in the positions so we can refund them, allocate_funds_per_position does not guarantee that all funds are used
+    CURRENT_DEPOSIT_LEFTOVER.save(
+        deps.storage,
+        &(
+            token0.amount - total_allocated.0,
+            token1.amount - total_allocated.1,
+        ),
     )?;
 
     Ok(Response::new()
-        .add_submessage(SubMsg::reply_on_success(
-            create_position_msg,
-            Replies::DepositCreatePosition as u64,
-        ))
+        .add_submessages(msgs??)
```

Comment:
Question - what is the mechanism when one of the sub msg failed here?

Reply:
All the submsges are reply on success, so if a submsg fails, we don't get a callback but get regular error behaviour instead, meaning the entire transaction is reverted